### ISSUE TITLE
Tracking Issue for `const fn` `type_id`

### ISSUE THREAD

**Issue by KodrAus** [rust-lang/rust#77125]
This is a tracking issue for using `TypeId::of` in constant contexts.
The feature gate for the issue is `#![feature(const_type_id)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Come up with a new scheme for collision resistant type ids
- [ ] Figure out any safeguards we need to prevent messing with the bits of type id directly
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

- How can we prevent abuses of transmuting type ids into integers at compile time (where we have fewer tools to deal with them)? See the discussion in #75923
- We need a new scheme for building type ids that's collision resistant. We'd like to do this _before_ stabilizing type ids in constant contexts

XXX --- list all the "unresolved questions" found in the RFC to ensure they are
not forgotten

### Implementation history

- Initial const implementation #47892
- Initial stabilization #72488
- Reverted stabilization #77083

**Comment by benkay86** [comment]
I'm not sure I totally understand the steps needed to re-stabilize `const fn TypeId::of()`. 

> - [ ] Come up with a new scheme for collision resistant type ids
> - We need a new scheme for building type ids that's collision resistant. We'd like to do this before stabilizing type ids in constant contexts

Does this affect the soundness of `const fn TypeId::of()`?  If we are really worried about collisions, then shouldn't the entirety of `TypeId` be unstable, and not just the `const` version?  Or are we worried that a future collision-resistant implementation of `TypeId` may not satisfy `const`, i.e. we want to keep open the possibility of a non-const future implementation?

> - [ ] Figure out any safeguards we need to prevent messing with the bits of type id directly
> - How can we prevent abuses of transmuting type ids into integers at compile time (where we have fewer tools to deal with them)? See the discussion in #75923

This seems impossible.  If someone wants to `transmute()` to the type of the private implementation (i.e. `u64`) in an `unsafe` code block, even though the documentation for `TypeId` explicitly says not to do this, can we really stop them?  (They can already do it at runtime, not sure what "tools" prevent this.)  Can we really stop anyone from doing unsafe things with `unsafe` code?

**Motivation:** It would be extremely useful to make `const` comparisons between `TypeId` in order to infer whether two types are equal at compile time.  Currently this only appears to be possible using specialization (#31844), which is unsound and will not be stabilized anytime soon.

A perhaps related concern (may need a separate tracking issue) is that `#derive[(PartialEq, Eq, ...)] struct TypeId` does not generate `const impl` for these traits (see #77695), so even if `const fn TypeId::of()` gets re-stabilized we still won't be able to compare types at compile time without writing a `const impl PartialEq for TypeId`, etc.

**Comment by jmeggitt** [comment]
# Can this be re-stabilized now?
@KodrAus Going through the comment that was the basis of reverting stabilization (https://github.com/rust-lang/rust/pull/75923#issuecomment-696676511):

## Part 1
> `TypeId` is one a few types (`mem::Discriminant` is the only other one I can think of) that's a thin wrapper for a private integer, at least right now. This is ripe for abuse no matter what we say/document it as.

Sure, but this suggests that retrieving a `TypeId` should be `unsafe` not `unstable`. `Any` is currently both stable and safe despite heavily relying on `const fn TypeId::of()`. If there are serious concerns that `const fn TypeId::of()` may result in collisions then `Any` should be unstable as well.

## Part 2
> I'm not saying we should cater to this, but rather avoid stabilizing even worse abuse-enabling tools like `const fn TypeId::of`.
Like I said, I am already aware of usecases where compile-time `transmute(TypeId::of::<T>())` is an uniquely useful tool and someone might not think twice before releasing a crate which uses it somewhere internally and becomes widely-used.

I think @benkay86 summed this up fairly well.
> If someone wants to `transmute()` to the type of the private implementation (i.e. `u64`) in an `unsafe` code block, even though the documentation for `TypeId` explicitly says not to do this, can we really stop them? (They can already do it at runtime, not sure what "tools" prevent this.) Can we really stop anyone from doing unsafe things with `unsafe` code?

## Part 3
And finally, @eddyb remarks that we could potentially re-stabilize it if we are unlikely to change it.
> If we decide we're unlikely to change the implementation of `TypeId`, then we can re-stabilize the `const fn` aspect, and live with the consequences. But if we do want to change the implementation away from a hash, we have to either do that first, or at least teach `miri` about it somehow so that it disallows using the value as a plain integer (for arithmetic or e.g. as an array type length).

It has been nearly 10 months since it was marked unstable and there have been no changes regarding this issue. At this point it seems like the implementation of `TypeId` is unlikely to change any time in the near future.

## TL;DR
Plus even if we assume all of these points are true,  **these concerns only apply to `intrinsics::type_id::<T>() -> u64` not `const fn TypeId::of::<T>() -> TypeId`**.

If anything, the heavy reliance on `const fn TypeId::of()` inside `Any` suggests that it is stable but could potentially be unsafe to use directly. If there are concerns that `TypeId` can be abused without unsafe code then `const fn TypeId::of()` should be marked as unsafe instead. However I do not think this is necessary since you would need to use `transmute()` to extract the raw type id integer from `TypeId`.

**`intrinsics::type_id::<T>()` should remain unstable as it provides access to the raw type id integer, but `const fn TypeId::of()` should be stabilized as it provides adequate protection.**

**Comment by eddyb** [comment]
> At this point it seems like the implementation of `TypeId` is unlikely to change any time in the near future.

Nobody tried AFAIK. 10 months is pretty short on the scale of Rust evolution, especially when you take into account the year 2020.

But also I believe it got stalled because of this potential breakage: https://github.com/rust-lang/rust/pull/75923#issuecomment-699096904

In order to guarantee code that `transmute`s to `u64` gets broken, we can't just replace `TypeId` with a pointer, but because we still want the 64-bit hash (for both "fast reject (`!=`)" and `Hash`, see https://github.com/rust-lang/rust/pull/75923#issuecomment-684917161), we can probably do this:
```rust
struct TypeId {
    hash: u64,
    // `str` can be replaced with a length-prefixed string to keep `TypeId`'s size down
    v0_mangled_type: &'static str,
}
impl PartialEq for TypeId {
    fn eq(&self, other: &TypeId) -> bool {
        self.hash == other.hash // fast reject (!=)
        && (
            self.v0_mangled_type as *const str == other.v0_mangled_type as *const str // fast accept (==)
            || self.v0_mangled_type == other.v0_mangled_type
        )
    }
}
```

That way, it's always more than 8 bytes, no matter the target pointer size, and so a `transmute` to `u64` will *always* error.

As for the linker concerns, if we take this specific approach, the only difference relying on their features would make is increasingly the situations in which the "fast accept (`==`)" path gets taken, i.e. an optimization.

I could try make a PR for this, but it would likely at least require https://github.com/rust-lang/rust/pull/87194 (to avoid breaking when unstable `const` generic features are used).

>  If there are concerns that `TypeId` can be abused without unsafe code then `const fn TypeId::of()` should be marked as unsafe instead.

But we don't *want* that. You can't change `TypeId` or `Any` to be `unsafe`, they're stable, and nearly-sound.

**Comment by programmerjake** [comment]
> ```rust
> struct TypeId {
>     hash: u64,
>     // `str` can be replaced with a length-prefixed string to keep `TypeId`'s size down
>     v0_mangled_type: &'static str,
> }
> impl PartialEq for TypeId {
>     fn eq(&self, other: &TypeId) -> bool {
>         self.hash == other.hash // fast reject (!=)
>         && (
>             self.v0_mangled_type as *const str == other.v0_mangled_type as *const str // fast accept (==)
>             || self.v0_mangled_type == other.v0_mangled_type
>         )
>     }
> }
> ```

Maybe it would be better to use this to shrink `TypeId` to only 16 bytes on 64-bit platforms:
```rust
use core::{ptr, slice};

#[repr(C)]
struct StrWithLen<T: ?Sized> {
    len: u16, // assume all TypeId strings are <64kB
    text: T,
}

pub struct TypeId {
    hash: u64,
    text: &'static StrWithLen<()>, // actually StrWithLen<str>
}

impl PartialEq for TypeId {
    fn eq(&self, rhs: &Self) -> bool {
        if self.hash != rhs.hash {
            false
        } else if ptr::eq(self.text, rhs.text) {
            true
        } else {
            unsafe {
                let lhs_text = slice::from_raw_parts(
                    &self.text.text as *const _ as *const u8,
                    self.text.len as usize,
                );
                let rhs_text = slice::from_raw_parts(
                    &rhs.text.text as *const _ as *const u8,
                    rhs.text.len as usize,
                );
                lhs_text == rhs_text
            }
        }
    }
}
```

**Comment by eddyb** [comment]
Yeah, that's what I meant by "length-prefixed". There's a "standard" way to do that using `extern { type }`, I just didn't want to type it in.

**Comment by eddyb** [comment]
Update: finally opened #95845 with a (v0) mangling-based `TypeId` representation (should unblock `const fn` `TypeId::of`, while at the same time making specific uses of `TypeId` incompatible with CTFE, for now).

**Comment by lcnr** [comment]
blocking the stabilization of this on #97156

**Comment by zakarumych** [comment]
Why would anyone use `transmute` to extract integer value from `TypeId`?
Use `Hash::hash` to avoid `unsafe`ty  :)

I know that `Hash` implementation for `TypeId` does not guarantee to emit inner value from `TypeId` unmodified, but transmuting opaque type is no better.

And why it is a problem anyway? I thought that transmuting arbitrary `u64` back to `TypeId` is.

**Comment by eddyb** [comment]
I'd say this is blocked on #99189 for now, if for no other reason than the "structural match" aspect.
(i.e. future-proofing `TypeId` by disallowing using `TypeId` constants as patterns and requiring `==`)

**Comment by istanbul-not-constantinople** [comment]
to work around #97156 and #99186, surely we can just add a new `const unsafe fn const_of() -> TypeId` to `TypeId` or even a new `ConstTypeId` type that uses the v0 mangled type. it seems the only problem with making `TypeId::of()` `const` is that it must be either non-`const` and safe or `const` and `unsafe` and since making a method `unsafe` is a breaking change, we cannot make it `const`. it seems obvious to me just to add a new `const unsafe fn` method and/or type that can explicitly warn the user of its soundness holes and potentially deprecate `TypeId::of()`.

**Comment by jmeggitt** [comment]
> I'd say this is blocked on #99189 for now, if for no other reason than the "structural match" aspect. (i.e. future-proofing `TypeId` by disallowing using `TypeId` constants as patterns and requiring `==`)

@eddyb What is the reasoning behind disallowing a structural match? While I can not name any specific reasons for its inclusion, I also can not think of any reasons to remove it either.

Take for example this code. It does not perform any complex pattern matching or require that a `match` statement be used, but it arguably makes the code a bit cleaner and easier to read than the alternative.
```rust
trait ConstTypeId {
    const ID: TypeId;
}

impl<T: 'static> ConstTypeId for T {
    const ID: TypeId = TypeId::of::<T>();
}

match foo.type_id {
    u16::ID | u32::ID | u64::ID => do_foo(&foo.data),
    Foo::ID | Bar::ID | Baz::ID => do_bar(&foo.data),
    _ => do_default(&foo.data),
}
```

This would generally serve the use case of "I should probably be using an `enum`, but there are a lot of variants and it is a pain to write it out". I'm partially joking, but there are some use cases where an `enum` can be more cumbersome such as when dealing with highly generic systems and associating resources with anonymous traits. That being said, I am unsure if those systems actually make use of use a `match` as to `==`.

**Comment by eddyb** [comment]
The problem is any *future* changes to `TypeId` that would require address-based comparison in a *custom* `PartialEq` implementation.

(See all previous discussion across this issue and my two closed PRs on the matter of non-trivial `TypeId` representations)

Honestly, our structural match rules are very loose and prone to bite us eventually.
What that is means that *in principle* no core/alloc/std type *with private and unguaranteed internals* should auto-derive `PartialEq, Eq` *and* provide a compile-time constructor, because that "locks in" the fact that the impls are auto-derived, i.e. that all fields themselves are `Eq`.

It's not a clear *independent* opt-in, for ergonomics reasons, but that also makes it *forwards-incompatable* to "let it happen".
(This applies to any crates but the standard library doesn't get to just release a new semver-incompat version, and editions offer only limited protection, and tend to have 3yr timeframes)

**Comment by yshui** [comment]
> This would generally serve the use case of "I should probably be using an enum, but there are a lot of variants and it is a pain to write it out".
> but there are some use cases where an enum can be more cumbersome such as when dealing with highly generic systems and associating resources with anonymous traits.

I actually am just writing such a system, and that's how I found out about this issue. And given the prospect that structural matching of `TypeId` could be gone in the future, I ended up using a set of macros to generate a big enum and a set of helper function to make working with it easier.

Not arguing for keeping it, just want to provide a data point.

**Comment by benkay86** [comment]
It seems `const fn type_id()` is going to take a long time to stabilize due to many concerns about soundness and corner cases of people doing something with the returned `TypeId` other than testing equality against another `TypeId`.  Would it be possible instead to make [`Any::is<T>(&self)`](https://doc.rust-lang.org/stable/std/any/trait.Any.html#method.is) `const`?  This way all the internals of `TypeId` stay inside the standard library.

**Comment by KodrAus** [comment]
I think a const-enabled `Any::is` that hid the specific equality mechanism would satisfy my use-case for const type ids too.

@eddyb did we have any lingering objection to being able to do this kind of type inspection at all in CTFE, or was it just the issue of structural match on `TypeId` that was the blocker?

**Comment by zakarumych** [comment]
I guess that most of the other use-cases would be covered by allowing `TypeId::of::<T>()` in const context, without allowing to do anything with it but assigning to a constant.

**Comment by Pr0methean** [comment]
Looks like TypeId suddenly stopped implementing StructuralEq. Why? I was previously able to use this function without a problem:
```
fn type_name_for_id(id: TypeId) -> &'static str {
    match id {
        PIXMAP_TYPE_ID => "Pixmap",
        MASK_TYPE_ID => "Mask",
        VEC_U8_TYPE_ID => "Vec<u8>",
        _ => "Unknown"
    }
}
```

**Comment by madsmtm** [comment]
Removing structural match for `TypeId` was done in https://github.com/rust-lang/rust/pull/103291, see the discussion above for reasoning behind it (shortened in https://github.com/rust-lang/rust/issues/77125#issuecomment-1221451165).

**Comment by Pr0methean** [comment]
Would it be possible to restore backward compatibility by having the compiler convert a `match` with TypeId constants into  an `else if` chain when there were only a handful of branches, and a lookup in a const `std::collections::HashMap` otherwise?

**Comment by oli-obk** [comment]
Yes, but that will require stabilization of `const_trait_impls` first, as otherwise we can't call PartialEq methods inside the const fn.

**Comment by KodrAus** [comment]
Most discussion about type ids in `const` contexts is happening here (naturally, since there's nothing to do with a type id except compare them), but there's also #101871 tracking using `==` in const contexts.

**Comment by briansmith** [comment]
I think https://github.com/rust-lang/rust/issues/10389 needs to be resolved first, becuase depending on the resolution, it may not be possible to calculate the TypeId at compile time; it might end up like C++ `std::type_index` that uses the address of the RTTI structure as the ID.

**Comment by oli-obk** [comment]
We can compute and equate type ids that are addresses just fine at compile time. You just can't inspect the address (so no sorting type ids, no hashing them, no printing them)

**Comment by tgross35** [comment]
> blocking the stabilization of this on #97156

This was recently marked fixed with https://github.com/rust-lang/rust/issues/97156

**Comment by avhz** [comment]
Any further updates on this ?

**Comment by bjorn3** [comment]
Still blocked on https://github.com/rust-lang/rust/issues/10389.

**Comment by MarcGuiselin** [comment]
It looks like while https://github.com/rust-lang/rust/issues/10389 was closed, there's still some ongoing discussion about the soundness/collision of type ids:

- https://github.com/rust-lang/rust/issues/129014
- https://github.com/rust-lang/rust/issues/129016
- https://github.com/rust-lang/rust/issues/129030

Is there anything else blocking?

There's still something that isn't clear to me:

> We need a new scheme for building type ids that's collision resistant. **We'd like to do this before stabilizing type ids in constant contexts.** 

I understand why https://github.com/rust-lang/rust/issues/101871 might be blocked due to collision issues but why this one? Ensuring `TypeId`s are collision resistant and being able to use `TypeId`s in const contexts seem like separate problems to me. The use case for this feature (such as in the case of https://github.com/bevyengine/bevy/issues/3365) of simply retrieving and including a `TypeId` as a field in a struct isn't unsound anymore afaik. Would it be okay to solve the collision issues after this feature and https://github.com/rust-lang/rust/issues/63084 have been stabilized?

**Comment by RalfJung** [comment]
The problem is that the scheme we end up using might involve pointers, but currently TypeId contains no pointers. If people transmute TypeId to integers in const, their code would stop compiling when we start having pointers in TypeId.

**Comment by bjoernager** [comment]
Is there any reason for allowing such casts to begin with? Skimming the docs, I see no mention of `TypeId` having a specific layout (in fact, it is explicitely stated that the type is opaque). Unless I am overlooking something important, I would say it would seem the most senseful to just declare that it has an unspecified representation, and that transmutations by extend would be undefined.

**Comment by RalfJung** [comment]
That's already the case, but it does not stop unsafe code from doing it anyway.

We could say "if you do that and your build breaks later, it's your own fault", but well, that often just doesn't work -- people will have their build broken on a Rust update because a dependency did something like this, and they will blame Rust, not the dependency.

**Comment by Jules-Bertholet** [comment]
I’m also worried about the interaction with `const fn` pointers/`dyn const`. Ideally, `const`-related errors in non-`const` contexts should be as rare as possible. To accomplish that, `const` types should smoothly degrade into their non-`const` versions in non-`const` contexts. I don’t know how that should best be modeled in `TypeId`, but it’s another reason to hold off on stabilizing this.

**Comment by bjoernager** [comment]
> That's already the case, but it does not stop unsafe code from doing it anyway.

Well, if it really is *undefined* behaviour, then I see no reason as to why it should be guaranteed to compile. I get the point that it currently does work (in practice, I guess), but whether it breaks at run-time, compile-time, or at all, is just an implementation detail IMHO.

**Comment by MarcGuiselin** [comment]
Thanks RalfJung, that is an interesting example. I didn't know about the TypeId layout changes and about it being possible to use transmute in const! (I can't even find a tracking issue for const_transmute.)

> I see no reason as to why it should be guaranteed to compile

That's kind of the point of stability though, to avoid your code breaking between rust updates. I understand the desire to hold off.

**Comment by bjoernager** [comment]
> That's kind of the point of stability though, to avoid your code breaking between rust updates.

If we have reached a point where definitively incorrect code is considered worthwhile enough to cater the language for and where rejecting such code is considered a breaking change, then we may as well no longer call it "undefined" behaviour if a specific outcome can always be safely relied upon.

**Comment by VorfeedCanal** [comment]
That's a tiny bit myopic. Sure, such code does exist… but **why** does it exist? It's not as if people write it to make life of Rust language developers harder.

The majority of people, when they reach out for the `type_id` is `const` want one thing only: to compare two `type_id`s (usually in `assert`, but sometimes to pick one of two branches with `if const`).

And AFAICS currently **the only** way to achieve that is via “transmute to integers”. You may complain about the people's desire to break the rules… but if something they very much want to do is to break the rules… they would break them anyway.

If I understand correctly the code in question, that @RalfJung fears to break is still nightly-only code and it's perfectly fine to break it… but this wouldn't make people happy if it would be broken with no suitable message – and that suitable message should, ideally, point to some supported-in-the-future way of doing the same thing.

So what's the solution for the most natural, obvious and desired feature of “`type_id` in `const`”? How can I compare two `type_id` in a `const` context to know if they are identical or not in a future-proof way?

**Comment by the8472** [comment]
> The problem is that the scheme we end up using might involve pointers, but currently TypeId contains no pointers. If people transmute TypeId to integers in const, their code would stop compiling when we start having pointers in TypeId.

Can we deliberately break this some way? E.g. by encoding the ints as pointers (and not telling const eval that they're really just integers)?

**Comment by RalfJung** [comment]
I can't think of a good way to do that. We'd need something like a magic `AllocId` that is assigned base address 0 when generating LLVM IR.

**Comment by benkay86** [comment]
Alternatively, the primary use case for `const TypeId` is _**comparing**_ `TypeId`, not _**storing**_ `TypeId`.  Could we stabilize a method something like `const fn compare(&self, other: &TypeId) -> bool` that performs the comparison without exposing the implementation details?  This way if the underlying type of `TypeId` changes in the future it would not be a breaking change.

The only problem I can see is if, in the future, the compiler can't prove two types are equal at compile time... but the compiler should always be able to do this, right?

**Comment by VorfeedCanal** [comment]
> Language design takes time, and this particular issue is unfortunately tied up with some long-standing non-trivial discussions.

Sure, but the only sane way forward is to do what was done with provenance: give users what they need or want in a safe way and **then** try to pressure them to stop using dirty hacks.

Without some way to do what they want to do they would find **another** unsafe and bug-ridden way… and then we would end up in the situation discussed.

> Could we stabilize a method something like `const fn compare(&self, other: &TypeId) -> bool` that performs the comparison without exposing the implementation details?

Such function is the only way forward IMO. I mean: what can you **actually do** to a `type_id` in a `const` block without something like this?

One may argue if that would be enough or if we also need hashing and some ordering… but without being able to even compare two `type_id`s in any way… what good they are for?

You can't exactly print them in a `const` function…

**Comment by RalfJung** [comment]
We have one option here that we did not have with provenance: don't let people use the problematic operation and type until we are confident in what we have.

**Comment by SafariMonkey** [comment]
I could be missing something, but `TypeId` appears to [implement `Eq`](https://doc.rust-lang.org/std/any/struct.TypeId.html#impl-Eq-for-TypeId) (and [even `Ord`](https://doc.rust-lang.org/std/any/struct.TypeId.html#impl-Ord-for-TypeId)), so I don't think you need to cast it to compare it unless you also e.g. want to stash it somewhere a `TypeId` can't otherwise go.

**Comment by bjorn3** [comment]
The Eq impl can't be used inside of a const fn as const trait impls aren't supported yet.

**Comment by benkay86** [comment]
@bjorn3 is quite right, and #67792 unfortunately appears to be quite far from stabilization.  However, that does not stop us from having a `const TypeId::is_same(&self, other: TypeId) -> bool` to implement the `const` comparison outside of a trait.  This would eventually become a vestigial method once `const` trait impls are stabilized.

**Comment by nvzqz** [comment]
If feasible, an alternative solution would be for the compiler to special-case `TypeId` to have `const` equality in the same way it does today for integers/char/bool.

**Comment by ArcaneNibble** [comment]
I would like to add a data point to this discussion -- I have some code (mostly as an experiment, not something "serious")  which specifically requires `TypeId::of` to be `const` but doesn't care about `const` equality or comparison at all.

Specifically, I've made some code which implements dynamic casting _between_ trait objects by using proc macros to build up run-time type information, [linkme](https://docs.rs/linkme/latest/linkme/) to put data tables together, and `unsafe` to meddle with vtable pointers. For every object type which implements traits which can be casted in this way, there is an array containing a list of (`TypeId::of::<dyn TraitA>`, _vtable pointer of_ `*const ObjType as *const dyn TraitA`), (`TypeId::of::<dyn TraitB>`, _vtable pointer of_ `*const ObjType as *const dyn TraitB`), ... entries. This requires `const` at compile-time in order to build up this array using `linkme` magic, but all of the actual casting action happens at runtime and so doesn't require `const` equality at all.

Because I want this to work on stable, and because this code already requires proc macros anyways, I've implemented something horrifying by using a `static AtomicU64` inside the proc macro crate and implementing a DIY, much-more-unsupported, and much-more-collision-prone `TypeId`.

(This code will always require unsafe, but it also incidentally currently uses `transmute` in order to substitute for `![feature(ptr_metadata)]`)

**Comment by zakarumych** [comment]
@ArcaneNibble you can use `TypeId::of` in const context. Specifically, you can grab the function pointer of a type `fn() -> TypeId`. Put it into const and static structures. And later evaluate for equality.

**Comment by ArcaneNibble** [comment]
> [@ArcaneNibble](https://github.com/ArcaneNibble) you can use `TypeId::of` in const context. Specifically, you can grab the function pointer of a type `fn() -> TypeId`. Put it into const and static structures. And later evaluate for equality.

Is this actually guaranteed to work? Attempting to do it this way gives `warning: function pointer comparisons do not produce meaningful results since their addresses are not guaranteed to be unique`, and I was assuming that Rust didn't make any guarantees or requirements like the one definition rule in C++.

(Or is this just supposed to be more reliable and less horrible than the `AtomicU64` hack?)

**Comment by zakarumych** [comment]
Exactly. It costs two function calls (you can compare function pointers first and call them if they are different though).
In exchange you get pointer-sized (smaller) type-id that you can acquire in const context.

**Comment by oli-obk** [comment]
Now that https://github.com/rust-lang/rust/pull/142789 has landed (well, as https://github.com/rust-lang/rust/pull/143696 after the first PR broke the master branch), We could stabilize const `TypeId::of` without stabilizing the `PartialEq` impl. It is not possible to transmute `TypeId` to `u128`. While there are still other things you could transmute it to, that code will break some day and is not very useful for actual type equality at compile-time.

**Comment by oli-obk** [comment]
@rust-lang/lang @rust-lang/libs-api what are your thoughts on stabilizing this without also stabilizing downcasting at compile-time?

**Comment by RalfJung** [comment]
One could transmute the last 8 bytes to u64 and compare them, even in `const fn`.

But... I guess if that kind of code breaks later we'll just say "you asked for this"?

... Hm that makes me wonder if we should just slap the same provenance on all the bytes? That's still simpler than your first proposal IIRC since the raw value of the bytes is the hash, so we don't need to delay the hash computation until codegen time.

**Comment by oli-obk** [comment]
All bytes of `TypeId` now have provenance. As long as we do not have a `wrapping` (out of bounds pointer supporting) `offset_of` method on pointers, it is impossible to extract any part of the hash from a TypeId. For example, the following program fails:

```rust
#![feature(const_type_id, const_trait_impl)]

use std::any::TypeId;

const _: isize = {
    let id = TypeId::of::<()>();
    let id: [*const u8; 2] = unsafe { std::mem::transmute(id) };
    unsafe { id[0].offset_from(id[1]) }
    //~^ ERROR: `ptr_offset_from` called on two different pointers where the memory range between them is not in-bounds of an allocation
};
```

And even if we added some sort of `wrapping_offset_from` that works in const eval, we could just adjust the TypeId internals to be hardened against that, too (by giving each segment its own provenance). Hopefully we'll have const traits on stable soon enough for that not to be something anyone wants to try.


This means TypeId is now entirely opaque at compile-time, so you can use it to copy it around and store it in a static place, but you can't use it for downcasting or even just TypeId comparisons, even with unsound operations like transmuting `TypeId`.

**Comment by RalfJung** [comment]
Agreed, I think TypeId is fully opaque at compile-time now. (And as a convenient side-effect, code running in Miri is hardened against type ID collisions. ;)

**Comment by traviscross** [comment]
> What are your thoughts on stabilizing this without also stabilizing downcasting at compile-time?

We talked about this in the lang team call.  We agreed it was still useful, as you had suggested, e.g. for people wanting to construct things containing a `TypeId` at compile-time (rather than storing a pointer to `TypeId::of::<T>`), and so we had appetite for this stabilization.

Please nominate the stabilization PR for us, and we'll FCP it there.

**Comment by oli-obk** [comment]
Stabilization PR is up: https://github.com/rust-lang/rust/pull/144133
